### PR TITLE
Connect EventDateTime block with new store

### PR DESCRIPTION
**Pull request by mitogh**
- Remove references to old store
- Remove references to Gutenberg Internals: withSelect, withDispatch
- Remove state from components and replace with props

### CODE REVIEW

File: src/modules/data/index.js [hunk 1]
```diff
@@ -6,12 +6,19 @@ import { composeWithDevTools } from 'redux-devtools-extension';
 import reducers from './reducers';
 import thunk from 'redux-thunk';
 
+import { dates, multiDay, allDay, endTime, startTime } from 'data/blocks/middlewares';
+
 const composeEnhancers = composeWithDevTools( {
```
Comment (paulmskim) [inline src/modules/data/index.js]: shouldn't we be using this for development only?

File: src/modules/data/blocks/middlewares/allday.js [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+const {
+	setStart,
+	setEnd,
+	setMultiDay,
+} = actions;
+
+export default ( { dispatch, getState } ) => ( next ) => ( action ) => {
+	next( action );
+
+	if ( action.type !== types.SET_ALL_DAY ) {
```
Comment (paulmskim) [inline src/modules/data/blocks/middlewares/allday.js]: I know that there's a check here for the action type, but If we're using other blocks, all the actions of those other blocks would flow through these middlewares (this one and the ones below). It seems like a waste of processing as well as semantically the wrong place to do this stuff. Could we do this processing when the specific action is processed instead?

File: src/modules/elements/time-picker/element.js [hunk 3]
```diff
@@ -151,77 +132,77 @@ export default class TimePicker extends Component {
...
+				{ this.currentLabel }
+				<Dashicon className="btn--icon" icon={ isOpen ? 'arrow-up' : 'arrow-down' } />
+			</button>
+		);
+	}
 
+	renderDropdownContent = ( { onToggle, isOpen, onClose } ) => {
+		this.onClose = onClose.bind( this );
+		return (
+			<ScrollTo>
```
Comment (paulmskim) [inline src/modules/elements/time-picker/element.js]: I'm not sure what the markup is for the dropdown content, but if it doesn't have an `aria-hidden` attribute, I think it should somewhere.

File: src/modules/blocks/event-datetime/block.js [hunk 4]
```diff
@@ -420,55 +454,38 @@ class EventDateTime extends Component {
...
+		cost: priceSelectors.getPrice( state ),
+	};
+};
+
+const mapDispatchToProps = ( dispatch ) => {
+	return {
+		...bindActionCreators( dateTimeActions, dispatch ),
+		...bindActionCreators( UIActions, dispatch ),
+		...bindActionCreators( priceActions, dispatch ),
+		setInitialState( attributes ) {
```
[the remark this answers is not in the document]
Comment (mitogh) [reply src/modules/blocks/event-datetime/block.js]: Yeah this one is called by a HOC attached to the block that populates the store when the component is mounted as the Gutenberg blocks has the data in the `attributes` property instead.

So we need a way to move those values into the store.

File: src/modules/data/blocks/datetime/actions.js [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+	type: types.SET_DATES,
+	meta: {
+		from,
+		to,
+	},
+} );
+
+export const setStart = ( start ) => ( {
+	type: types.SET_START_DATE_TIME,
+	payload: {
```
[the remark this answers is not in the document]
Comment (mitogh) [reply src/modules/data/blocks/datetime/actions.js]: Yeah at this point it seems like that however there are a couple of reasons to use this pattern: 

- Is used in most react apps and seems to be a good one: https://github.com/redux-utilities/flux-standard-action

- Provides a mechanism to standarise the way you set values or meta values to a reducer or middleware specially when there are more than one value.

File: src/modules/data/blocks/datetime/reducers.js [hunk 6]
```diff
@@ -1,17 +1,74 @@
...
-import separators from './separators/reducers';
+import { HALF_HOUR_IN_SECONDS } from 'editor/utils/time';
+import { roundTime, toDateTime } from 'editor/utils/moment';
+import { getSetting } from 'editor/settings';
+import * as types from './types';
+
+export const DEFAULT_STATE = {
+	start: toDateTime( roundTime( moment() ) ),
+	end: toDateTime( roundTime( moment() ).add( HALF_HOUR_IN_SECONDS, 'seconds' ) ),
+	dateTimeSeparator: getSetting( 'dateTimeSeparator', __( ' @ ', 'events-gutenberg' ) ),
```
[the remark this answers is not in the document]
Comment (mitogh) [reply src/modules/data/blocks/datetime/reducers.js]: Great will update to follow updates on the new ticket. 

Thanks for the heads up.

File: src/modules/data/blocks/middlewares/allday.js (hunk 2 above)
Comment (mitogh) [reply src/modules/data/blocks/middlewares/allday.js]: **Couple of follow up questions / comments**.

> It seems like a waste of processing as well

Well this one under the hood is basically a loop of middlewares that are executed or not dependent on the logic of each so it's only performant based on the amount of middlewares being executed and.

Take a look at: https://redux.js.org/faq/performance#how-well-does-redux-scale-in-terms-of-performance-and-architecture 

> semantically the wrong place to do this stuff.

Middlewares are actually the place to handle side effects or extend the basic functionally of reducers as the reducer should be a pure function. And the *All Day* functionally requires to handle external values and provide as well so middleware seems like the best place for this use case.

> Could we do this processing when the specific action is processed instead?

I'm not sure I get this one but this is the case at the moment as with the if we are not processing this for other actions.

File: src/modules/elements/time-picker/element.js (hunk 3 above)
Comment (mitogh) [reply src/modules/elements/time-picker/element.js]: Right! I'm not familiar with this dependency but I agree we should create a ticket so we can apply `aria` as well for other components that behave like buttons where some `aria` attributes are not defined as I think we have couple of more issues on A11y 👍

File: src/modules/data/blocks/datetime/selectors.js [hunk 7]
```diff
@@ -3,12 +3,39 @@
  */
 import { createSelector } from 'reselect';
 
-/**
- * Internal dependencies
- */
-import * as blockSelectors from 'data/blocks/selectors';
+export const datetimeSelector = ( state ) => state.blocks.datetime;
+
+export const getStart = createSelector(
```
[the remark this answers is not in the document]
Comment (mitogh) [reply src/modules/data/blocks/datetime/selectors.js]: Sure, couple of reasons behind this decision.

1. Consistency as most of the selectors are going to have the function siganature and we need a consistent mechanism to access data from the store.

2. Selectors are composable for more complex selection of data we need to compose multiple selectors (specially features like recurrence)

Let me know what you think

File: src/modules/data/blocks/datetime/reducers.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (mitogh) [reply src/modules/data/blocks/datetime/reducers.js]: Updated on #129

**Pull request merged**